lib/child-writer/index.js

```diff
@@ -170,14 +182,102 @@ exports.write = (image, drive, options) => {
     }
   }
 
+  // Timeout function that only runs if we exceed the specified timeout,
+  // terminating the server.
+  // @type {Timeout}
+  let terminationTimeout = null
+
+  // @type {Object} state
+  // @type {ChildProcess | undefined} [state.child]
+  // @type {Socket | undefined} [state.childSocket]
+  // @type {Boolean | undefined} [state.isChildRunning]
+  // @type {Boolean | undefined} [state.isChildSocketConnected]
+  const state = {}
+
+  /**
+   * @summary Finish the child writer
+   * @function
+   * @private
+   *
+   * @param {Number} code - exit code
+   * @returns {Void}
+   *
+   * @example
+   * finish(EXIT_CODES.SUCCESS)
+   */
+  const finish = (code) => {
+    terminateServer()
+    clearTimeout(terminationTimeout)
+
+    if (code === EXIT_CODES.CANCELLED) {
+      return emitter.emit('done', {
+        cancelled: true
+      })
+    }
+
+    // We shouldn't emit the `done` event manually here
+    // since the writer process will take care of it.
+    if (code === EXIT_CODES.SUCCESS || code === EXIT_CODES.VALIDATION_ERROR) {
+      return null
+    }
+
+    return emitError(new Error(`Child process exited with error code: ${code}`))
+  }
+
+  /**
+   * @summary Reset the timeout countdown
+   * @function
+   * @private
+   *
+   * @example
+   * resetTerminationTimeout()
+   */
+  const resetTerminationTimeout = () => {
+    clearTimeout(terminationTimeout)
+    terminationTimeout = setTimeout(() => {
+      const error = new Error('Write process timed out')
+      error.timeout = CHILD_WRITER_TIMEOUT_MS
+      error.isChildRunning = state.isChildRunning
+      error.childSocket = _.pick(state.childSocket, [
+        'connecting',
+        'destroyed',
+        'bytesRead',
+        'bytesWritten'
+      ])
+      error.childSocket.isConnected = state.isChildSocketConnected
+      error.child = _.pick(state.child, [ 'signalCode' ])
+      console.log(_.assign({}, error))
+
+      emitter.emit('timeout', error)
+
+      // Remove listeners so no more error popups are spawned
+      _.invoke(state.child, [ 'removeAllListeners' ])
+      _.invoke(state.child, [ 'kill' ])
```

Comment:
No need for `_.invoke()` here I believe

Reply:
It removes the need for a conditional checking whether `state.child` exists, because it'll otherwise error with `Cannot read property 'x' of undefined`

Reply:
Oh I see – the documentation mentions none of that, only "Invokes the method at path of object.", which is like "you can also just call the thing instead". I'd prefer to have an

```js
if( state.child ) {
  state.child.removeAllListeners()
  state.child.kill()
}
```

to make it obvious and clear what's going on here, and not swallow any errors.

Reply:
Sure. It's a sneaky property of Lodash, you can assume any method will just "fail" silently and return `undefined` if so.